rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+                    .offset(offset)
+                    .build();
+                PrimitiveArray::from(array_data)
+            }
+
+            /// Returns a `Buffer` holds all the values of this array.
+            pub fn values(&self) -> Buffer {
+                self.data.buffers()[0].copy()
+            }
+
+            /// Returns a raw pointer to the values of this array.
+            pub fn raw_values(&self) -> *const $native_ty {
+                unsafe { mem::transmute(self.raw_values.get().offset(self.data.offset() as isize)) }
+            }
+
+            /// Returns the primitive value at index `i`.
+            pub fn value(&self, i: i64) -> $native_ty {
+                unsafe { *(self.raw_values().offset(i as isize)) }
+            }
+
+            /// Determine the minimum value in the array
+            pub fn min(&self) -> Option<$native_ty> {
+                self.search(|a, b| a < b)
+            }
+
+            /// Determine the maximum value in the array
+            pub fn max(&self) -> Option<$native_ty> {
+                self.search(|a, b| a > b)
+            }
+
+            fn search<F>(&self, cmp: F) -> Option<$native_ty>
+            where
+                F: Fn($native_ty, $native_ty) -> bool,
+            {
+                let mut n: Option<$native_ty> = None;
+                let data = self.data();
+                for i in 0..data.length() {
+                    if data.is_null(i) {
+                        continue;
+                    }
+                    let m = self.value(i as i64);
+                    match n {
+                        None => n = Some(m),
+                        Some(nn) => if cmp(m, nn) {
+                            n = Some(m)
+                        },
+                    }
+                }
+                n
+            }
         }
-        ListArray::from(builder.finish())
-    }
+
+        /// Construct a primitive array from a vector. Should only be used for testing.
+        impl From<Vec<$native_ty>> for PrimitiveArray<$native_ty> {
+            fn from(data: Vec<$native_ty>) -> Self {
+                let array_data = ArrayData::builder($data_ty)
+                    .length(data.len() as i64)
+                    .add_buffer(Buffer::from(data.to_byte_slice()))
+                    .build();
+                PrimitiveArray::from(array_data)
+            }
+        }
+
+        impl From<Vec<Option<$native_ty>>> for PrimitiveArray<$native_ty> {
+            fn from(data: Vec<Option<$native_ty>>) -> Self {
+                const TY_SIZE: usize = mem::size_of::<$native_ty>();
+                const NULL: [u8; TY_SIZE] = [0; TY_SIZE];
+
+                let data_len = data.len() as i64;
+                let size = bit_util::round_upto_multiple_of_64(data_len) as usize;
+                let mut null_buffer = Vec::with_capacity(size);
+                unsafe {
+                    null_buffer.set_len(size);
+                }
```

Comment:
so now you have allocated a vector with undefined data it it, is that really what you want? I would rather prefer to have an explicit state here (like all-zero).